README.md
# TimeBar Table

A Flutter package for time-based data visualization. Create bar charts to represent data over time intervals with ease. Ideal for scheduling and event timelines.

![image](https://github.com/mihirrawal1399/timebar_table/assets/54273763/daabeee7-1517-41c1-aa55-a7aecc16c6f1)

## Features

- **Time-Based Visualization:** Easily represent data along a time axis with adjustable intervals.

- **Customization:** Tailor the appearance of your time bars, intervals, and more to suit your application's aesthetic.

- **Flexible Data Presentation:** Display various data elements simultaneously, enabling rich and informative timelines.

## Installation

To use this package, add it to your `pubspec.yaml` file:

```yaml
dependencies:
  timebar_table: ^0.0.1
```

Then, run flutter pub get to install the package.

## Usage
```dart
import 'package:flutter/material.dart';
import 'package:timebar_table/timebar_table.dart';

final DateTime tableStartTime = DateTime(2023, 9, 30, 8, 0);
final DateTime tableEndTime = DateTime(2023, 9, 30, 16, 0);
const Duration timeInterval = Duration(minutes: 30);
const double intervalWidth = 50.0;

// Define your data using a map where the keys are widgets and the values are lists of BarData.
final Map<Widget, List<BarData?>?> tableDataMap = {
  // Define your data here
};

final Widget emptyDataWidget = Container(
  width: intervalWidth,
  // Customize your empty data representation
);

void main() {
  runApp(MaterialApp(
    home: Scaffold(
      appBar: AppBar(
        title: const Text('TimeBarTable Example'),
      ),
      body: TimeBarTable(
        tableStartTime: tableStartTime,
        tableEndTime: tableEndTime,
        timeInterval: timeInterval,
        intervalWidth: intervalWidth,
        tableDataList: tableDataMap,
        emptyDataWidget: emptyDataWidget,
      ),
    ),
  ));
}
```

## Documentation

For detailed usage and customization options, consult the [API documentation](https://pub.dev/documentation/timebar_table/latest/).

## Example

To see how the package can be customized and integrated into your Flutter application, check out the included example.

## To-Do List

- [ ] Add support for Slivers and Custom Scroll Views.
- [ ] Modify readme.
- [ ] Implement Slivers and Custom Scroll Views as parameters for users.
- [ ] Provide additional customization parameters for elements.
- [ ] Introduce mapper functions for easily mapping custom objects for `tableDataList`.
- [ ] Maybe add code coverage.


## Contributions

Contributions are welcome! If you encounter any issues, have feature suggestions, or would like to contribute, please feel free to open an [issue](https://github.com/mihirrawal1399/timebar_table/issues) or submit a [pull request](https://github.com/mihirrawal1399/timebar_table/pulls).

## License
This package is released under the [MIT License](https://pub.dev/packages/timebar_table/license), allowing you to use, modify, and distribute it as needed.

## Author
Developed by [Mihir Rawal] - Contact me at [mihirrawal1399@gmail.com] for inquiries and support.


CHANGELOG.md
## 0.0.1

- Initial release of TimeBar Table Flutter package.
- Added support for displaying time-based bar charts.
- Features customizable components for data visualization.
- Enables the representation of data over specific time intervals.

## 0.0.2

- Fixed minor bugs in the package.
- Improved documentation.
- Updated example code.
- Updated Read me.

## 0.0.5

- Added TODO
- Modified Readme
- Added basic flutter Github action 

## 0.0.8

- changed description to match dart conventions
- import fixes for better pub dev score

## 0.1.0

- Added example screenshot for readme




test/timebar_table_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:timebar_table/timebar_table.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';

void main() {
  // testWidgets('Counter increments smoke test', (WidgetTester tester) async {
  //   // Build our app and trigger a frame.
  //   await tester.pumpWidget(const MyApp());

  //   // Verify that our counter starts at 0.
  //   expect(find.text('0'), findsOneWidget);
  //   expect(find.text('1'), findsNothing);

  //   // Tap the '+' icon and trigger a frame.
  //   await tester.tap(find.byIcon(Icons.add));
  //   await tester.pump();

  //   // Verify that our counter has incremented.
  //   expect(find.text('0'), findsNothing);
  //   expect(find.text('1'), findsOneWidget);
  // });
}


lib/src/time_row.dart
import 'package:flutter/material.dart';

class TimeRow extends StatelessWidget {
  const TimeRow({
    super.key,
    required this.tableStartTime,
    required this.tableEndTime,
    required this.timeInterval,
    required this.intervalWidth,
  });

  ///start time for the time range header
  final DateTime tableStartTime;

  ///end time for the time range header
  final DateTime tableEndTime;

  ///time intervals which interates from [tableStartTime] to [tableEndTime]
  final Duration timeInterval;

  ///width of the each [timeInterval]
  final double intervalWidth;

  @override
  Widget build(BuildContext context) {
    List<Widget> timeRowItems = [];

    // Loop through time intervals from tableStartTime to tableEndTime
    DateTime currentTime = tableStartTime;
    while (currentTime.isBefore(tableEndTime)) {
      final endTime = currentTime.add(timeInterval);
      final timeText =
          '${currentTime.hour}:${currentTime.minute.toString().padLeft(2, '0')} - ${endTime.hour}:${endTime.minute.toString().padLeft(2, '0')}';
      timeRowItems.add(
        TimeRowItem(
          intervalWidth: intervalWidth,
          text: timeText,
        ),
      );

      // Move to the next time interval
      currentTime = currentTime.add(timeInterval);
    }

    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
      children: [
        Container(
          width: 100,
          height: 100,
          alignment: Alignment.center,
          decoration: BoxDecoration(
            border: Border.all(color: Colors.black),
          ),
        ),
        ...timeRowItems,
      ],
    );
  }
}

class TimeRowItem extends StatelessWidget {
  final String text;
  final double intervalWidth;

  const TimeRowItem({
    required this.text,
    required this.intervalWidth,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      width: intervalWidth,
      height: 100,
      alignment: Alignment.center,
      decoration: BoxDecoration(
        border: Border.all(color: Colors.black),
      ),
      child: Text(
        text,
      ),
    );
  }
}


lib/timebar_table.dart
library timebar_table;

import 'package:flutter/material.dart';
import 'package:timebar_table/src/data_section.dart';

import 'src/table_column.dart';
import 'src/time_row.dart';

class TimeBarTable extends StatelessWidget {
  ///start time for the time range header
  final DateTime tableStartTime;

  ///end time for the time range header
  final DateTime tableEndTime;

  ///time intervals which interates from [tableStartTime] to [tableEndTime]
  final Duration timeInterval;

  ///width of the each [timeInterval]
  final double intervalWidth;

  ///no. of rows will be dependent on no. of keys of [tableDataList],
  ///and each row can have any list of [BarData] for given key.
  final Map<Widget, List<BarData?>?> tableDataList;

  ///custom empty widget to show if no BarData for perticular time lapse
  final Widget emptyDataWidget;

  const TimeBarTable({
    Key? key,
    required this.tableStartTime,
    required this.tableEndTime,
    required this.timeInterval,
    required this.intervalWidth,
    required this.tableDataList,
    this.emptyDataWidget = const SizedBox.shrink(),
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    //TODO: use slivers to implement custom scroll views making the main axises persistent scrollable.
    return Stack(
      children: [
        SingleChildScrollView(
          scrollDirection: Axis.horizontal,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              //the header row of the TimeBarTable which shows the time slots from [tableStartTime] to [tableEndTime] by [timeInterval]
              TimeRow(
                tableStartTime: tableStartTime,
                tableEndTime: tableEndTime,
                timeInterval: timeInterval,
                intervalWidth: intervalWidth,
              ),
              DataSection(
                tableDataMap: tableDataList,
                intervalWidth: intervalWidth,
                tableEndTime: tableEndTime,
                tableStartTime: tableStartTime,
                timeInterval: timeInterval,
                emptyDataWidget: emptyDataWidget,
              ),
            ],
          ),
        ),
        Positioned(
          top: 100,
          left: 0,
          child: TableColumn(
            widgets: tableDataList.keys.toList(),
          ),
        ),
      ],
    );
  }
}

class BarData {
  final Duration dataDuration;
  final DateTime dataStartTime;
  final Widget? dataWidget;

  const BarData({
    required this.dataDuration,
    required this.dataStartTime,
    this.dataWidget,
  });
}


lib/src/data_section.dart
import 'package:flutter/material.dart';

import '../timebar_table.dart';
import './firstwhereornull_extension.dart';

class DataSection extends StatelessWidget {
  final Map<Widget, List<BarData?>?> tableDataMap;

  ///width of the each [timeInterval]
  final double intervalWidth;

  ///start time for the time range header
  final DateTime tableStartTime;

  ///end time for the time range header
  final DateTime tableEndTime;

  ///time intervals which interates from [tableStartTime] to [tableEndTime]
  final Duration timeInterval;

  ///custom empty widget to show if no BarData for perticular time lapse
  final Widget emptyDataWidget;

  const DataSection({
    Key? key,
    required this.tableDataMap,
    required this.intervalWidth,
    required this.tableStartTime,
    required this.tableEndTime,
    required this.timeInterval,
    required this.emptyDataWidget,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        for (final entry in tableDataMap.entries) ...[
          DataRowBuilder(
            intervalWidth: intervalWidth,
            tableStartTime: tableStartTime,
            tableEndTime: tableEndTime,
            timeInterval: timeInterval,
            barDataList: entry.value,
            emptyDataWidget: emptyDataWidget,
          ),
        ],
      ],
    );
  }
}

class DataRowBuilder extends StatelessWidget {
  final List<BarData?>? barDataList;
  final DateTime tableStartTime;
  final DateTime tableEndTime;
  final Duration timeInterval;
  final Widget emptyDataWidget;
  final double intervalWidth;

  const DataRowBuilder({
    Key? key,
    this.barDataList,
    required this.tableStartTime,
    required this.tableEndTime,
    required this.timeInterval,
    required this.emptyDataWidget,
    required this.intervalWidth,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final List<Widget> rowWidgets = [];
    DateTime currentTime = tableStartTime;

    while (currentTime.isBefore(tableEndTime)) {
      if (barDataList != null) {
        // Check if there is a matching element in barDataList
        final matchingData = barDataList?.firstWhereOrNull((data) {
          return data != null &&
              (data.dataStartTime.isAtSameMomentAs(currentTime) ||
                  data.dataStartTime.isBefore(currentTime)) &&
              data.dataStartTime.add(data.dataDuration).isAfter(currentTime);
        });

        if (matchingData != null) {
          // Calculate the width based on the ratio of its duration to the time interval
          final durationRatio =
              matchingData.dataDuration.inMinutes / timeInterval.inMinutes;
          final barDataWidth = durationRatio * intervalWidth;

          rowWidgets.add(BarDataWidget(
            containerWidth: barDataWidth,
            customWidget: matchingData.dataWidget,
            emptyWidget: emptyDataWidget,
          ));
          currentTime =
              matchingData.dataStartTime.add(matchingData.dataDuration);
        } else {
          // Add an empty BarDataWidget
          rowWidgets.add(BarDataWidget(
            containerWidth: intervalWidth,
            emptyWidget: emptyDataWidget,
          ));
          currentTime = currentTime.add(timeInterval);
        }
      } else {
        rowWidgets.add(BarDataWidget(
          containerWidth: intervalWidth,
          emptyWidget: emptyDataWidget,
        ));
        currentTime = currentTime.add(timeInterval);
      }
    }

    return Row(
      children: [
        Container(
          width: 100,
          height: 50,
          decoration: BoxDecoration(
            border: Border.all(color: Colors.black),
          ),
        ),
        ...rowWidgets,
      ],
    );
  }
}

class BarDataWidget extends StatelessWidget {
  final double containerWidth;
  final Widget? customWidget;
  final Widget? emptyWidget;
  const BarDataWidget({
    super.key,
    required this.containerWidth,
    this.customWidget,
    this.emptyWidget,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      width: containerWidth,
      height: 50,
      alignment: Alignment.center,
      child: customWidget ?? emptyWidget,
    );
  }
}


lib/src/table_column.dart
import 'package:flutter/material.dart';

class TableColumn extends StatelessWidget {
  final List<Widget> widgets;
  const TableColumn({
    Key? key,
    required this.widgets,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        for (final widget in widgets) ...[
          TableColumnItem(widget: widget),
        ],
      ],
    );
  }
}

class TableColumnItem extends StatelessWidget {
  final Widget widget;

  const TableColumnItem({
    Key? key,
    required this.widget,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      width: 100,
      height: 50,
      alignment: Alignment.center,
      decoration: BoxDecoration(
        border: Border.all(color: Colors.black),
      ),
      child: widget,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:timebar_table/timebar_table.dart';

final DateTime tableStartTime =
    DateTime(2023, 9, 30, 8, 0); // Replace with your start time
final DateTime tableEndTime =
    DateTime(2023, 9, 30, 16, 0); // Replace with your end time
const Duration timeInterval =
    Duration(minutes: 30); // Replace with your time interval
const double intervalWidth = 50.0; // Replace with your interval width

// Define test data in the form of a map where the keys are widgets and the values are lists of BarData.
final Map<Widget, List<BarData?>?> tableDataMap = {
  const Text('Table 1'): [
    BarData(
      dataDuration: const Duration(minutes: 60),
      dataStartTime: DateTime(2023, 9, 30, 9, 0),
      dataWidget: Container(
        height: double.maxFinite,
        decoration: BoxDecoration(
          border: Border.all(),
          color: Colors.red,
        ),
        child: const Center(
          child: Text(
            "9am for 60mins",
            textAlign: TextAlign.center,
            style: TextStyle(
              color: Colors.white,
              fontWeight: FontWeight.w500,
            ),
          ),
        ),
      ),
    ),
    BarData(
      dataDuration: const Duration(minutes: 90),
      dataStartTime: DateTime(2023, 9, 30, 11, 0),
      dataWidget: Container(
        width: double.maxFinite,
        height: double.maxFinite,
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(8),
          color: Colors.blue,
        ),
        child: const Center(
          child: Text(
            "11am for 90mins",
            textAlign: TextAlign.center,
            style: TextStyle(
              color: Colors.red,
              fontWeight: FontWeight.w200,
            ),
          ),
        ),
      ),
    ),
  ],
  const Text('Table 2'): [
    BarData(
      dataDuration: const Duration(minutes: 120),
      dataStartTime: DateTime(2023, 9, 30, 10, 30),
      dataWidget: Container(
        width: double.maxFinite,
        height: double.maxFinite,
        decoration: BoxDecoration(
          border: Border.all(
            color: Colors.black,
            width: 4,
            strokeAlign: BorderSide.strokeAlignCenter,
          ),
          borderRadius: BorderRadius.circular(4),
          color: Colors.green,
        ),
        child: const Center(
          child: Text(
            "10am for 120mins",
            textAlign: TextAlign.center,
            style: TextStyle(
              color: Colors.black,
              fontWeight: FontWeight.w600,
            ),
          ),
        ),
      ),
    ),
  ],
  const Text('Table 3'): [
    BarData(
      dataDuration: const Duration(minutes: 60),
      dataStartTime: DateTime(2023, 9, 30, 8, 0),
      dataWidget: Container(
        width: double.maxFinite,
        height: double.maxFinite,
        decoration: BoxDecoration(
          border: Border.all(
            color: Colors.black,
            width: 2,
            strokeAlign: BorderSide.strokeAlignCenter,
          ),
          borderRadius: BorderRadius.circular(32),
          color: Colors.blue,
        ),
        child: const Center(
          child: Text(
            "8am for 60mins",
            textAlign: TextAlign.center,
            style: TextStyle(
              color: Colors.red,
              fontWeight: FontWeight.w400,
            ),
          ),
        ),
      ),
    ),
    BarData(
      dataDuration: const Duration(minutes: 120),
      dataStartTime: DateTime(2023, 9, 30, 11, 30),
      dataWidget: Container(
        width: double.maxFinite,
        height: double.maxFinite,
        decoration: BoxDecoration(
          border: Border.all(
            color: Colors.white,
            width: 4,
            strokeAlign: BorderSide.strokeAlignCenter,
          ),
          borderRadius: BorderRadius.circular(6),
          color: Colors.red,
        ),
        child: const Center(
          child: Text(
            "1130am for 120mins",
            textAlign: TextAlign.center,
            style: TextStyle(
              color: Colors.white,
              fontWeight: FontWeight.w400,
            ),
          ),
        ),
      ),
    ),
  ],
  const Text('Table 5'): [
    BarData(
      dataDuration: const Duration(minutes: 120),
      dataStartTime: DateTime(2023, 9, 30, 11, 0),
      dataWidget: Container(
        width: double.maxFinite,
        height: double.maxFinite,
        decoration: BoxDecoration(
          border: Border.all(
            color: Colors.blue,
            width: 4,
            strokeAlign: BorderSide.strokeAlignCenter,
          ),
          borderRadius: BorderRadius.circular(6),
          color: Colors.yellow,
        ),
        child: const Center(
          child: Text(
            "11am for 120mins",
            textAlign: TextAlign.center,
            style: TextStyle(
              color: Colors.black,
              fontWeight: FontWeight.w200,
            ),
          ),
        ),
      ),
    ),
    BarData(
      dataDuration: const Duration(minutes: 60),
      dataStartTime: DateTime(2023, 9, 30, 9, 0),
      dataWidget: Container(
        width: double.maxFinite,
        height: double.maxFinite,
        decoration: BoxDecoration(
          border: Border.all(
            color: Colors.blue,
            width: 4,
            strokeAlign: BorderSide.strokeAlignCenter,
          ),
          borderRadius: BorderRadius.circular(6),
          color: Colors.orange,
        ),
        child: const Center(
          child: Text(
            "9am for 60mins",
            textAlign: TextAlign.center,
            style: TextStyle(
              color: Colors.blue,
              fontWeight: FontWeight.w800,
            ),
          ),
        ),
      ),
    ),
  ],
  // Add more test data as needed
};

final Widget emptyDataWidget = Container(
  width: intervalWidth,
  decoration: BoxDecoration(
    border: Border.all(),
    color: Colors.red,
  ),
  padding: const EdgeInsets.all(2),
  child: Container(
    color: Colors.lightGreen,
  ),
);
void main() {
  runApp(MaterialApp(
    home: Scaffold(
      appBar: AppBar(
        title: const Text('TimeBarTable Example'),
      ),
      body: TimeBarTable(
        tableStartTime: tableStartTime,
        tableEndTime: tableEndTime,
        timeInterval: timeInterval,
        intervalWidth: intervalWidth,
        tableDataList: tableDataMap,
        emptyDataWidget: emptyDataWidget,
      ),
    ),
  ));
}
